core/src/main/java/org/elasticsearch/index/mapper/core/TypeParsers.java

```diff
@@ -308,7 +319,16 @@ public static void parseField(FieldMapper.Builder builder, String name, Map<Stri
                 // ignore for old indexes
                 iterator.remove();
             } else if (propName.equals("similarity")) {
-                builder.similarity(parserContext.getSimilarity(propNode.toString()));
+                String value = propNode.toString();
+                if (parserContext.indexVersionCreated().before(Version.V_3_0_0) &&
+                    "default".equals(value) && parserContext.getSimilarity(value) == null) {
+                    value = SimilarityService.DEFAULT_SIMILARITY;
+                }
+                SimilarityProvider similarityProvider = parserContext.getSimilarity(value);
+                if (similarityProvider == null) {
+                    throw new MapperParsingException("Unknown Similarity type [" + value + "] for [" + name + "]");
+                }
+                builder.similarity(similarityProvider);
```

Comment:
can you factor out the similarity parsing logic to a method to avoid duplication?